### PR TITLE
openthread-br: new package

### PR DESCRIPTION
**Pull request by stintel**
Maintainer: me
Compile tested: OpenWrt master r20360-832e7b8172 x86/64
Run tested: OpenWrt master r20360-832e7b8172 x86/64

Description:
This is the OpenThread Border Router. Tested with 2 nRF52480 sticks. One OpenWrt device configured as leader, then joined another OpenWrt device, the latter one ending up as router. Marked it as a draft for now. Test, comment, and ask away!

Leader:
```
# ot-ctl state
leader
Done

# ot-ctl neighbor table
| Role | RLOC16 | Age | Avg RSSI | Last RSSI |R|D|N| Extended MAC     |
+------+--------+-----+----------+-----------+-+-+-+------------------+
|   R  | 0x8c00 |   5 |      -76 |       -76 |1|1|1| c208cac4616521eb |

# ot-ctl router table
| ID | RLOC16 | Next Hop | Path Cost | LQ In | LQ Out | Age | Extended MAC     | Link |
+----+--------+----------+-----------+-------+--------+-----+------------------+------+
|  6 | 0x1800 |       63 |         0 |     0 |      0 |   0 | ea7f5177b595ccf7 |    0 |
| 35 | 0x8c00 |       63 |         0 |     3 |      3 |  10 | c208cac4616521eb |    1 |
```

Router:
```
# ot-ctl state
router
Done

# ot-ctl neighbor table
| Role | RLOC16 | Age | Avg RSSI | Last RSSI |R|D|N| Extended MAC     |
+------+--------+-----+----------+-----------+-+-+-+------------------+
|   R  | 0x1800 |  26 |      -75 |       -76 |1|1|1| ea7f5177b595ccf7 |
```

TODO:
- [ ]  ~~submit LuCI fixes upstream~~
- [x] decide on PKG_NAME (openthread-br, otbr-agent, ot-br-posix, something else?)
- [x] documentation (ran into a lot of WTFs, probably worth documenting them)
- [ ] ~~hotplug script to start/stop daemon on USB plug/unplug, based on serial, similar to what we've done for domoticz/zwave back in the day - and maybe also use this mechanism to generate the e.g. spinel+hdlc+uart:///dev/ttyACM0?uart-baudrate=460800 radio_url~~
- [x] avoid avahi/dbus dependency
- [x] figure out why settings are lost (I use CONFIG_TARGET_ROOTFS_PERSIST_VAR=y so /var/lib/thread is not on tmpfs) - https://github.com/openthread/openthread/blob/main/src/cli/README_DATASET.md probably holds the answer - find a clean way to load the dataset on boot

Not strictly related to otbr, but one of the WTFs is with the nRF52480 dongle. They seem to come with the u2f bootloader. To get it in mass storage device mode so you can just drop a file to flash a new firmware, you need to plug them in while pressing the reset button. Once I flashed the ot-rcp firmware this way, pressing the reset button while plugging in the dongle no longer toggles mass storage mode. Today I received more dongles, this behaviour is again confirmed. Initially I though I had flashed a wrong image and messed up the dongle. Turns out, to get the dongle in mass storage mode after flashing, you need to double press the reset button after plugging in the dongle.

Update:
The LuCI fixes seem no longer required, so I just dropped them. As for a hotplug handler, I decided to skip it. It can always be added later. I'd like to get this merged in its current form, the PR has been open long enough, and with the package available in our repos it will be much easier for other people to test and use.

### CODE REVIEW

**Comment by sgryphon** [comment]
I'd be keen on testing this... how do I get it onto my OpenWRT router?  Is there a beta package feed I need to configure, or manually point to a package?  (I have a few of the nRF52480, one already flashed with ot-ncp and tested with otbr-agent running in a RPi 4, but it would be nice to have it on my router)

**Comment by stintel** [comment]
> how do I get it onto my OpenWRT router?

You'll have to build your own image, with this PR manually applied to the local packages feed clone, or pointing feeds.conf directly to the openthread branch of my fork.

**Comment by sgryphon** [comment]
Hmmm. Is it possible if you already have built the .ipk to send it via email sgryphon+otbr@live.com ?

For the package name, if the only thing deployed is otbr-agent, then I suggest you just call it that (i.e. otbr-agent). This allows someone else to deploy otbr-web if they wanted (but I would recommend against that, as otbr-web as no authentication or security).

**Comment by stintel** [comment]
> Is it possible if you already have built the .ipk

Depends what target/subtarget you want to use this on...

**Comment by sgryphon** [comment]
Ah, I think I found the built packages:

* Click on Show all checks
* The click on Details for one of the builds
* Then click on Summary
* Scroll down to the Artifacts section
* Download arm_cortex-a9_vfpv3-d16... (I have a WRT1900AC, so mvebu/cortexa9)
* Unzip
* Find openthread-br_...ipk
* Use System > Software > Upload Package to install it (or you could transfer & do from command line)

Plug in the dongle, then reboot!

**Comment by sgryphon** [comment]
Oh, on the package name, I see it is already openthread-br in the OpenThread repo, so just leave it at that.

**Comment by sgryphon** [comment]
Hmmm, the error seems to be referring to basic_string char, traits and is maybe a compiler difference between the package and my OpenWRT (22.03.2, although the package would be main).

Reading at how the OpenWRT build chain works (for each target architecture it downloads all the compilers & source), this is maybe the case, i.e. different compilers used.

Do I maybe need to also run the corresponding beta / edge build from main?  Is that what you are doing, i.e. building everything together?

Alternatively, could I also build (maybe need to cherry pick the changes) against the 22.02.3 branch?

**Comment by ycc100** [comment]
1. Compile tested on : OpenWrt openwrt-22.03 <v22.03.2> (Qualcomm Atheros IPQ806X / NETGEAR Nighthawk X4S R7800 )

2. otbr-agent unable to start
-------------------------------------------------
root@OpenWrt:/# /etc/init.d/otbr-agent start
Tue Nov  8 06:49:43 2022 user.notice otbr-agent: validating agent section
root@OpenWrt:/# Tue Nov  8 06:49:44 2022 user.notice : [NOTE]-AGENT---: Backbone interface: eth1.1
Tue Nov  8 06:49:44 2022 user.notice otbr-agent[2852]: [NOTE]-AGENT---: Running 0.3.0
Tue Nov  8 06:49:44 2022 user.notice otbr-agent[2852]: [NOTE]-AGENT---: Thread version: 1.3.0
Tue Nov  8 06:49:44 2022 user.notice otbr-agent[2852]: [NOTE]-AGENT---: Thread interface: wpan0
Tue Nov  8 06:49:44 2022 user.notice otbr-agent[2852]: [NOTE]-AGENT---: Radio URL: spinel+hdlc+uart:///dev/ttyACM0
Tue Nov  8 06:49:44 2022 user.notice otbr-agent[2852]: [NOTE]-AGENT---: Radio URL: trel://eth1.1
Tue Nov  8 06:49:44 2022 user.notice otbr-agent[2852]: [NOTE]-ILS-----: Infra link selected: eth1.1
Tue Nov  8 06:49:44 2022 user.info otbr-agent[2852]: [INFO]-NCP-----: OpenThread log level changed to 4
Tue Nov  8 06:49:44 2022 user.info otbr-agent[2852]: 49d.22:12:20.180 [I] Platform------: RCP reset: RESET_EXTERNAL
Tue Nov  8 06:49:44 2022 user.info otbr-agent[2852]: 49d.22:12:20.185 [I] Platform------: RCP reset: RESET_POWER_ON
Tue Nov  8 06:49:44 2022 user.notice otbr-agent[2852]: 49d.22:12:20.196 [N] Platform------: RCP API Version: 6
Tue Nov  8 06:49:44 2022 user.info otbr-agent[2852]: 49d.22:12:20.197 [I] Platform------: Backbone interface is configured to eth1.1 (8)
Tue Nov  8 06:49:44 2022 user.crit otbr-agent[2852]: 00:00:00.032 [C] Platform------: exit(5): SetUp line 205, open, No such file or directory
Tue Nov  8 06:49:44 2022 daemon.info procd: Instance otbr-agent::instance1 s in a crash loop 8 crashes, 1 seconds since last crash
-------------------------------------------------

Deeply appreciate if anyone can help !

**Comment by stintel** [comment]
`ls /dev/ttyACM0`

**Comment by ycc100** [comment]
root@OpenWrt:/# ls -l /dev/ttyACM0 
crw-rw----    1 root     dialout   166,   0 Nov  9 01:25 /dev/ttyACM0
root@OpenWrt:/#

**Comment by stintel** [comment]
After checking the source, I suspect this comes from https://github.com/openthread/openthread/blob/main/src/posix/platform/daemon.cpp#L205
So find out the path of the lock file, then check if its parent path exists, etc. Maybe also enable verbose in the config file, might give more details.

**Comment by ycc100** [comment]
Solved. Root cause : The original R7800 has no /run/ directory.
Thanks a lot.

**Comment by LorbusChris** [comment]
Thanks for working on this!
Just fyi, there is a WIP PR upstream with some changes that one might want to import here: https://github.com/openthread/ot-br-posix/pull/1410
Also, it might make sense to upstream the contents of [999-fixes.patch](https://github.com/openwrt/packages/pull/19218/files#diff-f10ec4d9ba0accc99d01d83b804277c9caa0ceacacc9ead1f0e0eac69920cef2) instead of carrying that patch here.

**Comment by stintel** [comment]
> Just fyi, there is a WIP PR upstream with some changes that one might want to import here: [openthread/ot-br-posix#1410](https://github.com/openthread/ot-br-posix/pull/1410)

FYI, the changes in openthread/ot-br-posix#1410 (openthread/ot-br-posix#1673) would not have any impact on this PR, as we don't use any of those files. It does contain some useful changes, so thanks for pointing that out.


> Also, it might make sense to upstream the contents of [999-fixes.patch](https://github.com/openwrt/packages/pull/19218/files#diff-f10ec4d9ba0accc99d01d83b804277c9caa0ceacacc9ead1f0e0eac69920cef2) instead of carrying that patch here.

It's the very first item in the TODO of this PR.

**Comment by stintel** [comment]
Dumping some commands from `history` that I've ran to make it possible to actually transfer something from my desktop to an A64-Olinuxino that only has a Thread uplink to another A64-Olinuxino.

```
ot-ctl bbr enable
ot-ctl br enable
ot-ctl netdata publish route 0::/0 s high
```
At least something is actually working now!

```
openwrt-sunxi-cortexa53-olimex_a64-olinuxino-emmc-squashfs-sdcard.img.gz                                                 16% 2496KB   8.4KB/s   25:32 ETA 
```

**Comment by stintel** [comment]
Status update:

Upgrading EVE devices to the Matter firmware requires an Apple Homepod or Apple TV 4K with 128GB of RAM. Both are quite pricy if you were to get them just for this task. 

I've acquired an Apple TV and upgraded my EVE Motion to the Matter firmware. The Apple TV and the iPad running the EVE app where moved to an isolated network consisting of just an OpenWrt router, the iPad and the Apple TV. During the process, I've tcpdumped all traffic going over br-lan.

I am hoping to find a way to trick the EVE app into believing there is an Apple TV or Homepod Thread Border Router on the network, so that others can complete the Matter upgrade without the need for expensive Apple hardware.

As for the EVE Motion sensor with the Matter firmware, attempts to make it join my OpenWrt OpenThread network are ongoing using chip-tool.

For this, you need to decode the Matter QR code for the device to get the QR code payload. This can be done with zbarimg on Linux. You should have obtained this QR code from the EVE app during upgrade to Matter.

```
$ zbarimg eve_motion_89844940.jpg
QR-Code:MT:G0MB6WB-03MUXV65Y10
scanned 1 barcode symbols from 1 images in 0,03 seconds
```

You also need to obtain the Thread operational dataset.

```
ot-ctl dataset active -x
0e0800000000000123b435060004001fffe00208dead00beef00cafe0708fddead00beef000005100bcf552b6f84f21bd8593ae6d961d96c030b5468726541646c6576696f010216af0410000000000000000000000000000000000c0402a0f7f8000300001a
Done

```
Now you can try pairing the device. This uses BLE and thus requires a working bluetooth adapter with BLE support. I used a Linux laptop with an Intel AX210 for this.

0x6d01 is the node id. Pick a random 64 bit number. What I would probably end up doing in my production deployment is using ASCII to HEX to have meaningful prefixes followed by room number followed by device number, e.g.

M  O  T  I  O  N  3  1
6D 6F 74 69 6F 6E 03 01

```
./chip-tool pairing code-thread 0x6d01 hex:0e0800000000000123b435060004001fffe00208dead00beef00cafe0708fddead00beef000005100bcf552b6f84f21bd8593ae6d961d96c030b5468726541646c6576696f010216af0410000000000000000000000000000000000c0402a0f7f8000300001a MT:G0MB6WB-03MUXV65Y10
```

First problem: attestation verification

CHIP:CTL: Failed in verifying 'Attestation Information' command received from the device: err 101. Look at AttestationVerificationResult enum to understand the errors

err 101 = kPaaNotFound
https://github.com/project-chip/connectedhomeip/blob/master/src/credentials/attestation_verifier/DeviceAttestationVerifier.h#L35

Solution: point chip-tool to the production PAA root certificates
--paa-trust-store-path ...
https://github.com/project-chip/connectedhomeip/tree/master/credentials/production/paa-root-certs


Second problem: error on commissioning step 18 (FindOperational)
The EVE Motion joins the Thread Network, then chip-tool sends mDNS broadcast for 6C3AAEE0F9E563C2-0000000000006D01._matter._tcp.local. It either does not receive a response, or doesn't like the response. After a while, the device is removed from the Thread network. Have not found a solution for this so far.

**Comment by sgryphon** [comment]
> Upgrading EVE devices to the Matter firmware

Nice work. If it is helpful, I also have an Eve Energy device, that I used an Apple HomePod Mini to do the upgrade process to the Matter early access firmware.

If it helps, although it is not directly OpenWRT (although I do have one), I have so far got it partially working with Google Home:

* I then reset the device and commissioned it to Google Home via my Nest Hub Max. The WiFi, that both the HomePod and Hub Max are on is IPv6 only.
* After commissioning to Google Home, I can't actually get Google to control (on/off) the device (which I reported a bug to Eve/Google), but ...
* I can see the Matter device information (manufacturer, model, etc) in Google (only works if the Eve Energy is on).
* I was also able to get a pairing code from Google Home and then also add it to Apple Home, and then control the switch (on/off) via Apple Home.  The signal path is mobile --(IPv6 only wifi)--> HomePod --(IPv6 only wifi)--> Hub Max --(thread)--> Eve Energy.

Note that my Hub Max and HomePod Mini run separate Thread networks, and there is no way (yet) to share the credentials and join the network.

I also have a Raspberry Pi with nRF52840 dongle running OpenThread Border Router. Currently it is joined to the same Thread network as Google. I used a Nordic Thingy:53 which I commissioned to Google Home to grab the network credentials.

I have a second dongle plugged into my Linksys WRT 1900ac, which is running OpenWRT, but I haven't got OpenThread working on it yet. I had tried to set it up, but ran into some issues and have not gone back to try again yet. 

This may not be that useful, but it is at least proof that the Eve Energy, with the available firmware update, can commission to a non-Apple network (even if only partially working).

**Comment by stintel** [comment]
My Galaxy S21 Ultra 5G also detects the EVE Motion after the Matter upgrade. But to add it to SmartThings I need a SAMSUNG account, which I refuse to create.

As the EVE Motion is a battery powered device, I decided to also upgrade my EVE Energy to the Matter firmware, and try joining it with the chip-tool command. Unfortunately it fails on the same step as the EVE Motion. Not sure what to try next. Maybe I'll try with avahi again instead of mdnsresponder.

**Comment by stintel** [comment]
> Maybe I'll try with avahi again instead of mdnsresponder.

Same problem with avahi.

**Comment by stintel** [comment]
Looks like otbr-agent only registers the mDNS names after chip-tool already timed out:

```
[1674047811.907672][1813621:1813623] CHIP:DIS: OperationalSessionSetup[1:000000001234567A]: operational discovery failed: ../examples/chip-tool/third_party/connectedhomeip/src/lib/address_resolve/AddressResolve_DefaultImpl.cpp:109: CHIP Error 0x00000032: Timeout
```

```
date --date=@1674047811
wo 18 jan 2023 15:16:51 EET
```

```
Wed Jan 18 13:17:02 2023 user.info otbr-agent[2319]: [INFO]-MDNS----: Add subtype _CM for service D5144B42C0DE2241._matterc._udp
Wed Jan 18 13:17:02 2023 user.info otbr-agent[2319]: [INFO]-MDNS----: Add subtype _L2584 for service D5144B42C0DE2241._matterc._udp
Wed Jan 18 13:17:02 2023 user.info otbr-agent[2319]: [INFO]-MDNS----: Add subtype _S10 for service D5144B42C0DE2241._matterc._udp
Wed Jan 18 13:17:02 2023 user.info otbr-agent[2319]: [INFO]-MDNS----: Add subtype _T266 for service D5144B42C0DE2241._matterc._udp
Wed Jan 18 13:17:02 2023 user.info otbr-agent[2319]: [INFO]-MDNS----: Add subtype _V4874 for service D5144B42C0DE2241._matterc._udp
Wed Jan 18 13:17:02 2023 user.info otbr-agent[2319]: [INFO]-MDNS----: Commit avahi service D5144B42C0DE2241._matterc._udp
Wed Jan 18 13:17:03 2023 user.info otbr-agent[2319]: [INFO]-MDNS----: Avahi group (@0x7fbdccb710) is established
```

**Comment by stintel** [comment]
After increasing the mDNS timeout in the matter source from 45000 ms to 120000 ms, I am now hitting a Bluez disconnect. I guess I should figure out why otbr-agent is so slow with registering the mDNS names instead.

**Comment by stintel** [comment]
With Avahi I keep hitting this error:

```
Wed Jan 18 14:47:03 2023 daemon.debug avahi-daemon[3928]: sendmsg() to ff02::fb failed: Network unreachable
```

Running avahi with --no-drop-root doesn't change this. Some post suggests this is firewall related but the OpenWrt image I'm using doesn't have iptables, nor nftables, nor any netfilter kmods. Gonna build with CONFIG_KERNEL_IPV6_MROUTE disabled to see if that makes a difference.

**Comment by mips171** [comment]
Trying to test this. I have the ot-rcp firmware flashed onto my nRF52480, showing up at /dev/ttyACM0. Getting this error

```
Mon Apr 17 13:17:48 2023 user.notice otbr-agent: validating agent section
Mon Apr 17 13:17:48 2023 user.notice : [NOTE]-AGENT---: Backbone interface: eth0
Mon Apr 17 13:17:48 2023 user.notice otbr-agent[2479]: [NOTE]-AGENT---: Running 0.3.0-v21.02.7-1-gbb903b1ba1-dirty
Mon Apr 17 13:17:48 2023 user.notice otbr-agent[2479]: [NOTE]-AGENT---: Thread version: 1.3.0
Mon Apr 17 13:17:48 2023 user.notice otbr-agent[2479]: [NOTE]-AGENT---: Thread interface: wpan0
Mon Apr 17 13:17:48 2023 user.notice otbr-agent[2479]: [NOTE]-AGENT---: Radio URL: spinel+hdlc+uart:///dev/ttyACM0
Mon Apr 17 13:17:48 2023 user.notice otbr-agent[2479]: [NOTE]-AGENT---: Radio URL: trel://eth0
Mon Apr 17 13:17:48 2023 user.notice otbr-agent[2479]: [NOTE]-ILS-----: Infra link selected: eth0
Mon Apr 17 13:17:48 2023 user.info otbr-agent[2479]: [INFO]-NCP-----: OpenThread log level changed to 4
Mon Apr 17 13:17:48 2023 user.info otbr-agent[2479]: 49d.17:05:44.185 [I] Platform------: RCP reset: RESET_POWER_ON
Mon Apr 17 13:17:48 2023 user.info otbr-agent[2479]: 49d.17:05:44.197 [I] Platform------: Backbone interface is configured to eth0 (2)
Mon Apr 17 13:17:48 2023 user.crit otbr-agent[2479]: 49d.17:05:44.198 [C] Platform------: platformConfigureTunDevice() at netif.cpp:1697: No such file or directory
```

File: net/openthread-br/Makefile [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	+mdnsresponder
+endef
+
+define Package/openthread-br/description
+  A Thread border router for POSIX-based platforms.
+endef
+
+define Package/openthread-br/conffiles
+/etc/config/openthread
+/var/lib/thread
```
Comment (mips171) [inline net/openthread-br/Makefile:46]: I had to remove this to get it to compile on 21.02.7

File: net/openthread-br/Makefile [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	-DOTBR_TREL:BOOL=ON
+
+TARGET_CFLAGS += -DOPENTHREAD_POSIX_CONFIG_DAEMON_SOCKET_BASENAME=\\\"/var/run/openthread-%s\\\"
+
+define Package/openthread-br/install
+	$(INSTALL_DIR) \
+		$(1)/etc/config $(1)/etc/init.d \
+		$(1)/usr/lib/lua/luci/controller/admin \
+		$(1)/usr/lib/lua/luci/view/admin_thread $(1)/usr/sbin \
+		$(1)/var/lib/thread \
```
Comment (mips171) [inline net/openthread-br/Makefile:78]: `$(1)/var/lib/thread \`
I had to remove this to get it to compile on 21.02.7

Comment (stintel) [reply net/openthread-br/Makefile:78]: This is targeted at master. If you want to use it on anything else, you're on your own.

File: net/openthread-br/Makefile (hunk 1 above)
Comment (stintel) [reply net/openthread-br/Makefile:46]: This is targeted at master. If you want to use it on anything else, you're on your own.

**Comment by stintel** [comment]
Not sure why CI complains about PKG_MIRROR_HASH. Two people in OpenWrt confirmed they get the same hash as the one in the Makefile...

**Comment by stintel** [comment]
If nobody objects, I will merge this on 11/08/2023.

**Comment by feckert** [comment]
@stintel I can't find any LuCI source file in the pullrequest, but they are still mentioned in the Makefile. Did you forget to remove this from the Makefile or are the source files missing?

**Comment by stintel** [comment]
@feckert upstream has LuCI support

**Pull request merged**